haystack/components/joiners/branch.py

```diff
@@ -0,0 +1,147 @@
+    value coming to its single output, possibly distributing it to various other components.
+
+    `BranchJoiner` is fundamental to close loops in a pipeline, where the two branches it joins are the one
+    coming from the previous component, and the one coming back from a loop. For example, `BranchJoiner` could be used
+    to send data to a component evaluating errors. `BranchJoiner` would receive two connections, one to get the
+    original data, and another one to get modified data in case there was an error. In both cases, `BranchJoiner`
+    would send (or re-send in case of a loop) data to the component evaluating errors.
+
+    Another use case where there's the need of `BranchJoiner` is to reconcile multiple branches coming out of a decision
+    or classifier component. For example, in a RAG pipeline there might be a "query language classifier" component
+    sending the query to different retrievers, selecting one specifically according to the detected language. After the
+    retrieval step the pipeline would ideally continue with a `PromptBuilder`, and since we don't know in advance the
+    language of the query, all the retrievers should be ideally connected to the single `PromptBuilder`. Since the
+    `PromptBuilder` won't accept more than one connection in input, we would connect all the retrievers to a
+    `BranchJoiner` component and reconcile them in a single output that can be connected to the `PromptBuilder`
+    downstream.
+
+    Usage example:
+
+    ```python
+    import json
+    from typing import List
+
+    from haystack import Pipeline
+    from haystack.components.converters import OutputAdapter
+    from haystack.components.generators.chat import OpenAIChatGenerator
+    from haystack.components.joiners import BranchJoiner
+    from haystack.components.validators import JsonSchemaValidator
+    from haystack.dataclasses import ChatMessage
+
+    person_schema = {
+        "type": "object",
+        "properties": {
+            "first_name": {"type": "string", "pattern": "^[A-Z][a-z]+$"},
+            "last_name": {"type": "string", "pattern": "^[A-Z][a-z]+$"},
+            "nationality": {"type": "string", "enum": ["Italian", "Portuguese", "American"]},
+        },
+        "required": ["first_name", "last_name", "nationality"]
+    }
+
+    # Initialize a pipeline
+    pipe = Pipeline()
+
+    # Add components to the pipeline
+    pipe.add_component('mx', BranchJoiner(List[ChatMessage]))
+    pipe.add_component('fc_llm', OpenAIChatGenerator(model="gpt-3.5-turbo-0125"))
+    pipe.add_component('validator', JsonSchemaValidator(json_schema=person_schema))
+    pipe.add_component('adapter', OutputAdapter("{{chat_message}}", List[ChatMessage])),
+    # And connect them
+    pipe.connect("adapter", "mx")
+    pipe.connect("mx", "fc_llm")
+    pipe.connect("fc_llm.replies", "validator.messages")
+    pipe.connect("validator.validation_error", "mx")
```

Comment:
let's change `mx` to `branch_joiner`, if possible

Reply:
And also perhaps refer your first description to an example below?